ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -2089,7 +2100,34 @@ private void finalizeMetadataChunk() {
               passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
               passedInBlobProperties.isEncrypted(), passedInBlobProperties.getExternalAssetTag(),
               passedInBlobProperties.getContentEncoding(), passedInBlobProperties.getFilename(), null);
-      if (isStitchOperation() || getNumDataChunks() > 1) {
+
+      if (isS3MultipartPartUpload()) {
+        // close the request as the single blob. we don't generate the composite blob.
+        blobId = (BlobId) indexToChunkIdsAndChunkSizes.get(0).getFirst();
+        state = ChunkState.Complete;
+        setOperationCompleted();
+
+        // now generate the data chunk id list as the S3 ETag
+        // values returned are in the right order as TreeMap returns them in key-order.
+        List<Pair<StoreKey, Long>> orderedChunkIdList = new ArrayList<>(indexToChunkIdsAndChunkSizes.values());
+        logger.info("S3 5MB : finalizeMetadataChunk orderedChunkIdList is {} {}", indexToChunkIdsAndChunkSizes,
+            restRequest);
+
+        JSONObject jsonObject = new JSONObject();
+        for (Pair<StoreKey, Long> blobAndSize : orderedChunkIdList) {
+          JSONObject oneObject =
+              new JSONObject().put("blob", blobAndSize.getFirst()).put("size", blobAndSize.getSecond());
+          jsonObject.accumulate("chunks", oneObject);
+        }
+        jsonObject.put("ttl", ""); // placeholder
+        jsonObject.put("reservedMetadataId", ""); // placeholder
+        jsonObject.put("session", ""); // placeholder
+        String result = jsonObject.toString();
+        logger.info("S3 5MB : finalizeMetadataChunk generated json {}", result);
+
+        // use the S3_BLOB_LIST to pass the value
+        restRequest.setArg(S3_BLOB_LIST, result);
```

Comment:
Ideally we let S3 code path to generate the ETag, not here. 
Anyway to pass the dataChunk raw data to the caller?